src/middlewared/middlewared/plugins/cluster_linux/ctdb_private.py

```diff
@@ -0,0 +1,241 @@
+
+        # the documentation clearly states that before
+        # adding a private peer, the cluster must be
+        # healthy. This requires running a command that
+        # is expecting the ctdb daemon to be started.
+        # If this is the first private peer being added
+        # then the ctdb daemon isn't going to be started
+        # which means we can't check if the cluster is
+        # healthy. So we do the following:
+        #   1. if the ctdb shared volume private ip file exists
+        #       then assume that this isn't the first peer
+        #       being added to the cluster and check the ctdb
+        #       daemon for the cluster health.
+        #   2. elif the ctdb shared volume private ip doesnt
+        #       exist then assume this is the first peer being
+        #       added to the cluster and skip the cluster health
+        #       check.
+        ctdb_file = pathlib.Path(CTDBConfig.GM_PRI_IP_FILE.value)
+        etc_file = pathlib.Path(CTDBConfig.ETC_PRI_IP_FILE.value)
+
+        if ctdb_file.exists():
+            if self.middleware.call_sync('service.started', 'ctdb'):
+                if not self.middleware.call_sync('ctdb.general.healthy'):
+                    self.logger.error('ctdb cluster is not healthy')
+                    return False
+
+        # make sure the private ip file exists
+        try:
+            ctdb_file.touch(exist_ok=True)
+        except Exception:
+            self.logger.error('Failed creating %s', str(ctdb_file), exc_info=True)
+            return False
+
+        # we need to make sure that the local etc private ip file
+        # is symlinked to the ctdb shared volume private ip file
+        symlink_it = delete_it = False
+        if etc_file.exists():
+            if not etc_file.is_symlink():
+                # delete it since we're symlinking it
+                delete_it = True
+            else:
+                # means it's a symlink but not to the ctdb
+                # shared volume private ip file
+                if not etc_file.resolve() == ctdb_file:
+                    delete_it = True
+        else:
+            symlink_it = True
+
+        if delete_it:
+            try:
+                etc_file.unlink()
+            except Exception:
+                self.logger.error('Failed deleting %s', str(etc_file), exc_info=True)
+                return False
+
+        if symlink_it:
+            try:
+                etc_file.symlink_to(ctdb_file)
+            except Exception:
+                self.logger.error('Failed symlinking %s to %s', str(etc_file), str(ctdb_file), exc_info=True)
+                return False
+
+        # ctdb documentation is _VERY_ explicit about
+        # how this file should be modified
+        if not delete:
+            # in the case of adding a node, it _MUST_ be
+            # added to the end of the file always
+            with open(PRI_IP_FILE, 'a') as f:
+                f.write(pri_ip + '\n')
+                f.flush()
```

Comment:
flush() is probably not necessary, since a close() from the context manager should imply it